### ISSUE TITLE
F# Record Serialization

### ISSUE THREAD

**Issue by jhugard** [ctaggart/froto#38]
@ctaggart has requested the ability to serialize/deserialize F# Record types, rather than .NET Class Types.

**Comment by jhugard** [comment]
Proposed protobuf serializable Record definition

```
    type ETest =
        | Nada = 0
        | One = 1
        | Two = 2

    let ETestDefault = ETest.One  // NOTE: Non-zero default is only supported in Proto2
    
    type InnerSample = {
        id : int32
        name : string
        option : bool
        test : ETest
        packedFixed32 : uint32 list
        repeatedInt32 : int32 list
        }
        with
            static member Default = {
                id=0
                name=""
                option=false
                test=ETest.Nada
                packedFixed32=[]
                repeatedInt32=[]
            }

            member m.Serialize zcb                      = zcb |> serialize m
            member m.Deserialize zcb                    = zcb |> deserialize m
            member m.SerializeLengthDelimited zcb       = zcb |> serializeLengthDelimited m
            member m.DeserializeLengthDelimited zcb     = zcb |> deserializeLengthDelimited m

            member m.SerializeWithUnknowns unks zcb                 = zcb |> serializeWithUnknowns m unks
            member m.DeserializeWithUnknowns zcb                    = zcb |> deserializeWithUnknowns m
            member m.SerializeLengthDelimitedWithUnknowns unks zcb  = zcb |> serializeLengthDelimitedWithUnknowns m unks
            member m.DeserializeLengthDelimitedWithUnknowns zcb     = zcb |> deserializeLengthDelimitedWithUnknowns m
                
            member m.Serializer zcb =
                (m.id            |> dehydrateVarint 1) >>
                (m.name          |> dehydrateString 2) >>
                (m.option        |> dehydrateBool 3) >>
                (m.test          |> dehydrateDefaultedVarint ETestDefault 4) >>
                (m.packedFixed32 |> dehydratePackedFixed32 5) >>
                (m.repeatedInt32 |> dehydrateRepeated dehydrateVarint 6)
                <| zcb

            member m.DecoderRing =
                [
                    1, fun m rawField -> { m with id = rawField |> hydrateInt32 }
                    2, fun m rawField -> { m with name = rawField |> hydrateString }
                    3, fun m rawField -> { m with option = rawField |> hydrateBool }
                    4, fun m rawField -> { m with test = rawField |> hydrateEnum }
                    5, fun m rawField -> { m with packedFixed32 = rawField |> hydratePackedFixed32 }
                    6, fun m rawField -> { m with repeatedInt32 = (rawField |> hydrateInt32) :: m.repeatedInt32 }
                ]
                |> Map.ofList

            static member DecodeFixup m =
                { m with
                    packedFixed32 = List.rev m.packedFixed32
                    repeatedInt32 = List.rev m.repeatedInt32 }
```

**Comment by ctaggart** [comment]
This makes me really happy to see this!

**Comment by jhugard** [comment]
Fixed in #39
